File: sections/08_binary_search_answer/problems/c_minimum_capacity/solution.py
"""Reference solution for C. Minimum Capacity."""

import sys


def days_needed(weights: list[int], capacity: int) -> int:
    days = 1
    cur = 0
    for w in weights:
        if cur + w <= capacity:
            cur += w
        else:
            days += 1
            cur = w
    return days


def solve_case(weights: list[int], days: int) -> int:
    lo = max(weights)
    hi = sum(weights)
    while lo < hi:
        mid = (lo + hi) // 2
        if days_needed(weights, mid) <= days:
            hi = mid
        else:
            lo = mid + 1
    return lo


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        days = data[idx + 1]
        idx += 2
        weights = data[idx:idx + n]
        idx += n
        out.append(str(solve_case(weights, days)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/08_binary_search_answer/problems/c_minimum_capacity/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int days_needed(const vector<long long> &weights, long long capacity) {
    int days = 1;
    long long cur = 0;
    for (long long w : weights) {
        if (cur + w <= capacity) {
            cur += w;
        } else {
            ++days;
            cur = w;
        }
    }
    return days;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n, d;
        cin >> n >> d;
        vector<long long> weights(n);
        long long lo = 0;
        long long hi = 0;
        for (long long &w : weights) {
            cin >> w;
            lo = max(lo, w);
            hi += w;
        }
        while (lo < hi) {
            long long mid = lo + (hi - lo) / 2;
            if (days_needed(weights, mid) <= d) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        cout << lo << '\n';
    }

    return 0;
}

